### ISSUE TITLE
List comprehensions are awkward

### ISSUE THREAD

**Issue by chrisdone** [chrisdone/hindent#174]
For all styles hindent currently presents list comprehensions like
this:

``` haskell
[Foob.Mordr 0 s ba p 0 q | s <- ["a", "b"]
                         , ba <- [Foob.Dude, Foob.Pie]
                         , p <- [20000, 20001, 20002, 20003]
                         , q <- [100, 200]]
```

This is awkward and uses space poorly. But nobody uses list
comprehensions so nobody really minded until now. It's bugging me
enough that I decided to fix it.

**Comment by chrisdone** [comment]
Fixed in 12d3b51. 

Now it formats by default as:

``` haskell
λ> testAll "x = [x*y*z|y <- 1,x <- 2,z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog]"
-- fundamental:
x = 
  [x * y * z
  |y <- 1
  ,x <- 2
  ,z <- 3 * 5 + 2 + foo
                      (bar
                         mu)
                      bob
                      bill
                      zot
                      dog]

-- chris-done:
x = [x * y * z|y <- 1,x <- 2,z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog]

-- johan-tibell:
x = 
    [ x * y * z
    | y <- 1 
    , x <- 2 
    , z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog ]

-- gibiansky:
x = [ x * y * z
    | y <- 1 
    , x <- 2 
    , z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog ]

-- cramer:
x = [ x * y * z
    | y <- 1 
    , x <- 2 
    , z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog ]
```

``` haskell
λ> testAll "x = [x*y*z|y <- 1,x <- 2,z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog,p <- z]"
-- fundamental:
x = 
  [x * y * z
  |y <- 1
  ,x <- 2
  ,z <- 3 * 5 + 2 + foo
                      (bar
                         mu)
                      bob
                      bill
                      zot
                      dog
  ,p <- z]

-- chris-done:
x = 
  [x * y * z
  |y <- 1
  ,x <- 2
  ,z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog
  ,p <- z]

-- johan-tibell:
x = 
    [ x * y * z
    | y <- 1 
    , x <- 2 
    , z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog 
    , p <- z ]

-- gibiansky:
x = [ x * y * z
    | y <- 1 
    , x <- 2 
    , z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog 
    , p <- z ]

-- cramer:
x = [ x * y * z
    | y <- 1 
    , x <- 2 
    , z <- 3 * 5 + 2 + foo (bar mu) bob bill zot dog 
    , p <- z ]
```

Pinging @ocharles @gibiansky @ennocramer.

**Comment by michalrus** [comment]
One more thing, that you might want to consider. When hindenting this (using your style):

```haskell
xs = [x | x <- [1,2,3]]
```

… we’re getting:

```haskell
xs = [x|x <- [1,2,3]]
```

… and `[x|` is then treated as a beginning of a quasi-quote:

```
hindent-reformat-region-as-is: hindent: Unexpected end of input while lexing quasi-quoter
```
